src/datachain/lib/udf.py

```diff
@@ -517,26 +529,40 @@ def run(
         def _prepare_rows(udf_inputs) -> "abc.Generator[Sequence[Any], None, None]":
             with safe_closing(udf_inputs):
                 for row in udf_inputs:
-                    yield self._prepare_row(
+                    yield self._prepare_row_and_id(
                         row, udf_fields, catalog, cache, download_cb
                     )
 
-        def _process_row(row):
-            with safe_closing(self.process_safe(row)) as result_objs:
-                for result_obj in result_objs:
-                    udf_output = self._flatten_row(result_obj)
-                    yield dict(zip(self.signal_names, udf_output, strict=False))
-
+        # Prepare and prefetch inputs (ID is included and harmlessly skipped by
+        # prefetch)
         prepared_inputs = _prepare_rows(udf_inputs)
         prepared_inputs = _prefetch_inputs(
             prepared_inputs,
             self.prefetch,
             download_cb=download_cb,
             remove_prefetched=bool(self.prefetch) and not cache,
         )
+
+        # Process rows, extracting ID for checkpoint tracking
         with closing(prepared_inputs):
-            for row in prepared_inputs:
-                yield _process_row(row)
+            for row_id, *udf_args in prepared_inputs:
+                # TODO: Fix limitation where inputs yielding nothing are not tracked in
+                # processed table. Currently, if process() yields nothing for an input,
+                # that input's sys__id is never added to the processed table, causing it
+                # to be re-processed on checkpoint recovery. Solution: yield a marker
+                # row with sys__input_id when process() yields nothing, then filter
+                # these marker rows before inserting to output table.
+                output_batch = []
+                with safe_closing(self.process_safe(udf_args)) as result_objs:
+                    for result_obj in result_objs:
+                        udf_output = self._flatten_row(result_obj)
+                        # Include sys__input_id to track which input generated this
+                        # output
+                        output_batch.append(
+                            {"sys__input_id": row_id}
```

Comment:
Lets not use this name (`sys__input_id`) since `Sys` schema does not have this field (`input_id`)?